src/language_features/semantic_highlighting.rs

```diff
@@ -0,0 +1,82 @@
+use crate::context::*;
+use crate::position::*;
+use crate::types::*;
+use itertools::Itertools;
+use jsonrpc_core::Params;
+use lsp_types::*;
+
+pub fn semantic_highlighting_notification(params: Params, ctx: &mut Context) {
+    let params = params.parse::<SemanticHighlightingParams>();
+    let params = params.unwrap();
+    let path = params.text_document.uri.to_file_path().unwrap();
+    let buffile = path.to_str().unwrap();
+    let document = match ctx.documents.get(buffile) {
+        Some(document) => document,
+        None => return,
+    };
+    let meta = match ctx.meta_for_buffer(buffile.to_string()) {
+        Some(meta) => meta,
+        None => return,
+    };
+    let scopes = ctx
+        .capabilities
+        .as_ref()
+        .unwrap()
+        .semantic_highlighting
+        .as_ref()
+        .unwrap()
+        .scopes
+        .as_ref()
+        .expect("Server sent semantic highlight notification without setting capability");
+    let offset_encoding = ctx.offset_encoding.to_owned();
+    let ranges = params
+        .lines
+        .iter()
+        .map(|info| {
```

Comment:
nit:  [flat_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) might fit there.

Reply:
Ah yes, ofc, I had some issue and replaced it with this and forgot to change it back when that wasn't it

Reply:
ah no, it had something todo with some scoping errors i didn't fully understand, and this was the easiest way for me to fix it

Reply:
This is weird, I don't understand either why does it complain about `line` lifetime as `u64` could be just copied. However, we can use `move` and also take `offset_encoding` by reference instead of cloning it (performed implicitly by `to_owned`):

```
        .flat_map(|info| {
            let line: u64 = info.line as u64;
            let offset_encoding = &ctx.offset_encoding;
            info.tokens.iter().map(move |t| {
                let scope = scopes
                    .get(t.scope as usize)
                    .expect("Semantic highlighting token sent for out-of-range scope");
                let face = get_face_for_scope(scope);
                let range = Range {
                    start: Position::new(line, t.character.into()),
                    end: Position::new(line, (t.character + u32::from(t.length)).into()),
                };
                format!(
                    "{}|{}",
                    lsp_range_to_kakoune(&range, &document.text, offset_encoding),
                    face
                )
            })
        })
```